### PR TITLE
Stop hibernated room

### PR DESCRIPTION
**Pull request by michalwski**
This PR introduces deep hibernation of room's process.
The optimisation works only for persistent rooms for which the process can be stopped and recreated later on demand.

Goes on top of #1086

### CODE REVIEW

File: apps/ejabberd/src/mod_muc.erl [hunk 1]
```diff
@@ -481,43 +530,67 @@ route_to_room(Room, {From,To,Packet} = Routed, #state{host=Host} = State) ->
...
+                    Packet, ?ERRT_NOT_ALLOWED(Lang, ErrText)),
             ejabberd_router:route(To, From, Err)
     end.
 
+route_packet_to_nonexistent_room(Room, From, To, Packet,
+                                 #state{server_host = ServerHost,
+                                        host = Host,
+                                        access = Access} = State) ->
+
+    case ?BACKEND:restore_room(ServerHost, Host, Room) of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Don't use macros (like BACKEND on line 575) as module names.

Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Remove the dynamic function call on line 575. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 2]
```diff
@@ -670,15 +677,45 @@ handle_info(process_room_queue, normal_state = StateName, StateData) ->
...
+handle_info(_Info, StateName, #state{hibernate_timeout = Timeout} = StateData) ->
+    {next_state, StateName, StateData, Timeout}.
+
+maybe_stop_persistent_room(RoomName, true, State) ->
+    do_stop_persistent_room(RoomName, State);
+maybe_stop_persistent_room(RoomName, _, State) ->
+    stop_if_only_owner_is_online(RoomName, count_users(State), State).
+
+stop_if_only_owner_is_online(RoomName, 1, #state{users = Users, jid = RoomJID} = State) ->
+    [{LJID, #user{jid = LastUser, nick = Nick}}] = ?DICT:to_list(Users),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Remove the dynamic function call on line 699. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/mod_muc.erl [hunk 3]
```diff
@@ -481,43 +530,67 @@ route_to_room(Room, {From,To,Packet} = Routed, #state{host=Host} = State) ->
...
+            {ok, Pid} = mod_muc_room:start(Host, ServerHost, Access,
+                                           Room, HistorySize,
+                                           RoomShaper, HttpAuthPool, Opts),
+            {_, _, Nick} = jid:to_lower(To),
+            register_room(Host, Room, Pid),
+            mod_muc_room:route(Pid, From, Nick, Packet)
+    end.
 
 -spec route_by_nick(room(), from_to_packet(), state()) -> 'ok' | pid().
 route_by_nick(<<>>, {_,_,Packet} = Routed, State) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Elvis**:
> Missing space right "," on line 596

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 4]
```diff
@@ -3781,6 +3823,311 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+    ok;
+maybe_configure(Owner, RoomName, Opts) ->
+    maybe_set_subject(proplists:get_value(subject, Opts), Owner, RoomName),
+    Cfg = [opt_to_room_config(Opt) || Opt <- Opts],
+    Form = stanza_configuration_form(RoomName, lists:flatten(Cfg)),
+    escalus:send(Owner, Form),
+
+    Result = escalus:wait_for_stanza(Owner),
+    escalus:assert(is_iq_result, Result),
+    escalus:assert(is_stanza_from, [<<RoomName/binary,"@muc.localhost">>], Result).
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Missing space right "," on line 4043

File: apps/ejabberd/src/mod_muc_room.erl [hunk 5]
```diff
@@ -690,18 +727,7 @@ terminate(Reason, _StateName, StateData) ->
...
-                    children = [#xmlel{name = <<"x">>,
-                                       attrs = [{<<"xmlns">>, ?NS_MUC_USER}],
-                                       children = [#xmlel{name = <<"item">>,
-                                                          attrs = ItemAttrs,
-                                                          children = [ReasonEl]},
-                                                   #xmlel{name = <<"status">>,
-                                                          attrs = [{<<"code">>, <<"332">>}]}
-                                                  ]}]},
+    Packet = unavailable_presence(ItemAttrs, ReasonT),
     ?DICT:fold(
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Don't use macros (like DICT on line 731) as module names.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 6]
```diff
@@ -3781,6 +3823,311 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+
+room_with_only_owner_is_stopped(Config) ->
+    RoomName = fresh_room_name(),
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        {ok, _, Pid} = given_fresh_room_is_hibernated(
+                         Alice, RoomName, [{persistentroom, true}]),
+        true = wait_for_room_to_be_stopped(Pid, timer:seconds(8)),
+
+        Unavailable = escalus:wait_for_stanza(Alice),
+	    escalus:assert(is_presence_with_type, [<<"unavailable">>], Unavailable)
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3982 has a tab at column 0.

File: apps/ejabberd/src/mod_muc_room.erl (hunk 5 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Remove the dynamic function call on line 731. Only modules that define callbacks should make dynamic calls.

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 7]
```diff
@@ -3781,6 +3823,311 @@ deny_creation_of_http_password_protected_room_service_unavailable(Config) ->
...
+        Stanza = stanza_set_affiliations(RoomName, [{escalus_client:short_jid(Bob), <<"member">>}]),
+        escalus:send(Alice, Stanza),
+        escalus:assert(is_iq_result, escalus:wait_for_stanza(Alice)),
+        is_invitation(escalus:wait_for_stanza(Bob)),
+
+        leave_room(RoomName, Alice),
+
+        true = wait_for_room_to_be_stopped(Pid, timer:seconds(8)),
+
+        Stanza2 = stanza_set_affiliations(RoomName, [{escalus_client:short_jid(Kate), <<"member">>}]),
```
Comment (gadgetci) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: According to **Elvis**:
> Line 3952 is too long:         Stanza2 = stanza_set_affiliations(RoomName, [{escalus_client:short_jid(Kate), <<"member">>}]),.

File: apps/ejabberd/src/mod_muc_room.erl (hunk 2 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Don't use macros (like DICT on line 699) as module names.

File: apps/ejabberd/src/mod_muc.erl [hunk 8]
```diff
@@ -481,43 +530,67 @@ route_to_room(Room, {From,To,Packet} = Routed, #state{host=Host} = State) ->
 
 -spec route_to_nonexistent_room(room(), from_to_packet(), state()) -> 'ok'.
 route_to_nonexistent_room(Room, {From, To, Packet},
-                          #state{host=Host} = State) ->
+                          #state{host = Host} = State) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Compiler**:
> Warning: variable 'Host' is unused

**Comment by gadgetci** [comment]
**compiler** failed with exit status: 1

File: apps/ejabberd/src/mod_muc.erl [hunk 9]
```diff
@@ -481,43 +530,67 @@ route_to_room(Room, {From,To,Packet} = Routed, #state{host=Host} = State) ->
...
-                    ErrText = <<"Room creation is denied by service policy">>,
-                    Err = jlib:make_error_reply(
-                            Packet, ?ERRT_NOT_ALLOWED(Lang, ErrText)),
-                    ejabberd_router:route(To, From, Err)
-            end;
+            route_presence_to_nonexistent_room(Room, From, To, Packet, State);
         _ ->
-            Lang = xml:get_attr_s(<<"xml:lang">>, Attrs),
-            ErrText = <<"Conference room does not exist">>,
+            route_packet_to_nonexistent_room(Room, From, To, Packet, State)
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Compiler**:
> function route_packet_to_nonexistent_room/5 undefined

File: apps/ejabberd/src/mod_muc.erl (hunk 1 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc.erl]: According to **Compiler**:
> undefined macro 'BACKEND'

File: apps/ejabberd/src/mod_muc_room.erl [hunk 10]
```diff
@@ -670,15 +679,45 @@ handle_info(process_room_queue, normal_state = StateName, StateData) ->
...
+handle_info(_Info, StateName, #state{hibernate_timeout = Timeout} = StateData) ->
+    {next_state, StateName, StateData, Timeout}.
+
+maybe_stop_persistent_room(RoomName, true, State) ->
+    do_stop_persistent_room(RoomName, State);
+maybe_stop_persistent_room(RoomName, _, State) ->
+    stop_if_only_owner_is_online(RoomName, count_users(State), State).
+
+stop_if_only_owner_is_online(RoomName, 1, #state{users = Users, jid = RoomJID} = State) ->
+    [{LJID, #user{jid = LastUser, nick = Nick}}] = ?DICT:to_list(Users),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Remove the dynamic function call on line 701. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 11]
```diff
@@ -690,18 +729,7 @@ terminate(Reason, _StateName, StateData) ->
...
-                    children = [#xmlel{name = <<"x">>,
-                                       attrs = [{<<"xmlns">>, ?NS_MUC_USER}],
-                                       children = [#xmlel{name = <<"item">>,
-                                                          attrs = ItemAttrs,
-                                                          children = [ReasonEl]},
-                                                   #xmlel{name = <<"status">>,
-                                                          attrs = [{<<"code">>, <<"332">>}]}
-                                                  ]}]},
+    Packet = unavailable_presence(ItemAttrs, ReasonT),
     ?DICT:fold(
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Remove the dynamic function call on line 733. Only modules that define callbacks should make dynamic calls.

File: apps/ejabberd/src/mod_muc_room.erl (hunk 10 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Don't use macros (like DICT on line 701) as module names.

File: apps/ejabberd/src/mod_muc_room.erl (hunk 11 above)
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Don't use macros (like DICT on line 733) as module names.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 12]
```diff
@@ -2190,16 +2213,16 @@ erase_matched_users_dict(LJID, Users, Sessions) ->
         {U, S, <<>>} ->
             FF = fun(J, #user{nick=Nick}, {Us, Ss}) ->
                      case J of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> The expression on line 2215 and column 22 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 13]
```diff
@@ -2776,7 +2799,7 @@ search_role(Role, StateData) ->
     lists:filter(
       fun({_, #user{role = R}}) ->
           Role == R
-      end, ?DICT:to_list(StateData#state.users)).
+      end, dict:to_list(StateData#state.users)).
 
 
 -spec search_affiliation(mod_muc:affiliation(), state()) -> [{_,_}].
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Missing space right "," on line 2805

File: apps/ejabberd/src/mod_muc_room.erl [hunk 14]
```diff
@@ -4746,7 +4756,7 @@ route_nick_iq(#routed_nick_iq{allow_query = true, online = {true, _, _}, jid = f
         From, Err);
 route_nick_iq(#routed_nick_iq{allow_query = true, online = {true, NewId, FromFull},
     jid = ToJID, packet = Packet, stanza = StanzaId}, StateData) ->
-    {ok, #user{nick = FromNick}} = ?DICT:find(jid:to_lower(FromFull),
+    {ok, #user{nick = FromNick}} = dict:find(jid:to_lower(FromFull),
         StateData#state.users),
     {ToJID2, Packet2} = handle_iq_vcard(FromFull, ToJID,
         StanzaId, NewId,Packet),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Missing space right "," on line 4762

File: apps/ejabberd/src/mod_muc_room.erl [hunk 15]
```diff
@@ -2261,7 +2284,7 @@ send_new_presence(NJID, StateData) ->
 
 -spec send_new_presence(ejabberd:jid(), binary(), state()) -> 'ok'.
 send_new_presence(NJID, Reason, StateData) ->
-    send_new_presence_to(NJID, Reason, ?DICT:to_list(StateData#state.users), StateData).
+    send_new_presence_to(NJID, Reason, dict:to_list(StateData#state.users), StateData).
 
 
 -spec send_new_presence_to(ejabberd:jid(), binary(), [{ejabberd:jid(), #user{}}], state()) -> 'ok'.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> The spec in line 2290 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 16]
```diff
@@ -2789,7 +2812,7 @@ search_affiliation(Affiliation, StateData) when is_atom(Affiliation) ->
           _ ->
               Affiliation == A
           end
-      end, ?DICT:to_list(StateData#state.affiliations)).
+      end, dict:to_list(StateData#state.affiliations)).
 
 
 -spec process_admin_items_set(ejabberd:jid(), [jlib:xmlel(),...], ejabberd:lang(),
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Missing space right "," on line 2818

File: apps/ejabberd/src/mod_muc_room.erl [hunk 17]
```diff
@@ -3993,7 +4003,7 @@ rfield(Label, Var, Val, Lang) ->
 
 -spec iq_disco_info_extras(ejabberd:lang(), state()) -> [jlib:xmlel(),...].
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Missing space right "," on line 4004

File: apps/ejabberd/src/mod_muc_room.erl [hunk 18]
```diff
@@ -899,7 +941,7 @@ is_allowed_nonparticipant(JID, StateData) ->
 -spec get_participant_data(ejabberd:simple_jid() | ejabberd:jid(),
                            state()) -> {_,_}.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Missing space right "," on line 942

File: apps/ejabberd/src/mod_muc_room.erl [hunk 19]
```diff
@@ -2219,15 +2242,15 @@ update_matched_users_dict(F, LJID, Users) ->
         {U, S, <<>>} ->
             FF = fun(J, User, Us) ->
                      case J of
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> The expression on line 2244 and column 22 is nested beyond the maximum level of 3.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 20]
```diff
@@ -2261,7 +2270,7 @@ send_new_presence(NJID, StateData) ->
 
 -spec send_new_presence(ejabberd:jid(), binary(), state()) -> 'ok'.
 send_new_presence(NJID, Reason, StateData) ->
-    send_new_presence_to(NJID, Reason, ?DICT:to_list(StateData#state.users), StateData).
+    send_new_presence_to(NJID, Reason, dict:to_list(StateData#state.users), StateData).
 
 
 -spec send_new_presence_to(ejabberd:jid(), binary(), [{ejabberd:jid(), #user{}}], state()) -> 'ok'.
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> The spec in line 2276 uses a record, please define a type for the record and use that instead.

File: apps/ejabberd/src/mod_muc_room.erl [hunk 21]
```diff
@@ -2789,11 +2798,11 @@ search_affiliation(Affiliation, StateData) when is_atom(Affiliation) ->
           _ ->
               Affiliation == A
           end
-      end, ?DICT:to_list(StateData#state.affiliations)).
+      end, dict:to_list(StateData#state.affiliations)).
 
 
--spec process_admin_items_set(ejabberd:jid(), [jlib:xmlel(),...], ejabberd:lang(),
-                        state()) -> {'error', jlib:xmlel()} | {'result',[],state()}.
+-spec process_admin_items_set(ejabberd:jid(), [jlib:xmlel(),...], ejabberd:lang(), state()) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_muc_room.erl]: According to **Elvis**:
> Missing space right "," on line 2804

File: test/ejabberd_tests/tests/muc_SUITE.erl [hunk 22]
```diff
@@ -82,10 +82,24 @@ all() -> [
         {group, owner_no_parallel},
         {group, room_management},
         {group, http_auth_no_server},
-        {group, http_auth}
+        {group, http_auth},
+          {group, hibernation}
```
Comment (fenek) [inline test/ejabberd_tests/tests/muc_SUITE.erl]: Indent

**Pull request merged**